zcash_primitives/src/merkle_tree/incremental.rs

```diff
@@ -0,0 +1,341 @@
+
+impl HashSer for MerkleCrhOrchardOutput {
+    fn read<R: Read>(mut reader: R) -> io::Result<Self>
+    where
+        Self: Sized,
+    {
+        let mut repr = [0u8; 32];
+        reader.read_exact(&mut repr)?;
+        <Option<_>>::from(Self::from_bytes(&repr)).ok_or_else(|| {
+            io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "Non-canonical encoding of Pallas base field value.",
+            )
+        })
+    }
+
+    fn write<W: Write>(&self, mut writer: W) -> io::Result<()> {
+        writer.write_all(&self.to_bytes())
+    }
+}
+
+pub fn write_nonempty_frontier_v1<H: HashSer, W: Write>(
+    mut writer: W,
+    frontier: &NonEmptyFrontier<H>,
+) -> io::Result<()> {
+    writer.write_u64::<LittleEndian>(<u64>::from(frontier.position()))?;
+    match frontier.leaf() {
+        Leaf::Left(a) => {
+            a.write(&mut writer)?;
+            Optional::write(&mut writer, &None, |w, n: &H| n.write(w))?;
+        }
+        Leaf::Right(a, b) => {
+            a.write(&mut writer)?;
+            Optional::write(&mut writer, &Some(b), |w, n| n.write(w))?;
+        }
+    }
+    Vector::write(&mut writer, &frontier.ommers(), |w, e| e.write(w))?;
+
+    Ok(())
+}
+
+#[allow(clippy::redundant_closure)]
+pub fn read_nonempty_frontier_v1<H: HashSer + Clone, R: Read>(
+    mut reader: R,
+) -> io::Result<NonEmptyFrontier<H>> {
+    let position = reader
+        .read_u64::<LittleEndian>()
+        .map(|v| <Position>::from(v as usize))?;
+    let left = H::read(&mut reader)?;
+    let right = Optional::read(&mut reader, |r| H::read(r))?;
+
+    let leaf = right.map_or_else(
+        || Leaf::Left(left.clone()),
+        |r| Leaf::Right(left.clone(), r),
+    );
+    let ommers = Vector::read(&mut reader, |r| H::read(r))?;
+
+    NonEmptyFrontier::from_parts(position, leaf, ommers).map_err(|err| {
+        io::Error::new(
+            io::ErrorKind::InvalidInput,
+            format!("Parsing resulted in an invalid Merkle frontier: {:?}", err),
+        )
+    })
+}
+
+pub fn write_frontier_v1<H: HashSer, W: Write>(
+    mut writer: W,
+    frontier: &Frontier<H, 32>,
+) -> io::Result<()> {
+    match frontier.value() {
+        None => {
+            writer.write_u8(0)?;
+        }
+        Some(f) => {
+            writer.write_u8(1)?;
+            write_nonempty_frontier_v1(&mut writer, f)?;
+        }
+    }
+
+    Ok(())
+}
+
+#[allow(clippy::redundant_closure)]
+pub fn read_frontier_v1<H: HashSer + Clone, R: Read>(mut reader: R) -> io::Result<Frontier<H, 32>> {
+    let is_empty = reader.read_u8()? == 0;
+    if is_empty {
+        Ok(Frontier::empty())
+    } else {
+        read_nonempty_frontier_v1(reader).and_then(|f| {
+            Frontier::try_from(f).map_err(|err| {
+                io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    format!("Parsing resulted in an invalid Merkle frontier: {:?}", err),
+                )
+            })
+        })
+    }
```

Comment:
This currently allows for a malleable format: `Some` can be represented by any non-zero byte. It also means the format is not round-trip serializable, as `write_frontier_v1` does not reproduce the leading byte.

Use `serialize::Optional::read` instead:
```suggestion
pub fn read_frontier_v1<H: HashSer + Clone, R: Read>(reader: R) -> io::Result<Frontier<H, 32>> {
    match Optional::read(reader, read_nonempty_frontier_v1)? {
        None => Ok(Frontier::empty()),
        Some(f) => Frontier::try_from(f).map_err(|err| {
            io::Error::new(
                io::ErrorKind::InvalidInput,
                format!("Parsing resulted in an invalid Merkle frontier: {:?}", err),
            )
        }),
    }
```